[ENG-3834] Institution SSO Email Domain Update Script

cslzchen opened the pull request:
## Purpose

Support institutions that have requested a institutional wide email domain change.

## Changes

N/A

## Dev / DevOps Notes

* How to run the script

```
  -h, --help         show this help message and exit
  -i ID, --id ID     the institution of which affiliated users' eligible email
                     is to be updated
  -s SRC, --src SRC  the email domain that are eligible for update
  -d DST, --dst DST  the email domain that is to be added
```

```
docker-compose run --rm web python3 -m  -i <instn_id> -s <src_domain> -d <dst_domain>
```

* Example local logs for adding `<username>@test.osf.io` to `cos` users who has `<username>@osf.io`.

```
[__main__]  INFO: Action aborted since email [longze@test.osf.io] exists for a different user user [wbg6v]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[__main__]  INFO: Successfully added email [roger@test.osf.io] to user [wbg6v]!
[__main__]  INFO: Action skipped since email [naxin@test.osf.io] exists for the same user [mtf5k]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[__main__]  INFO: Successfully added email [longze+affiliation@test.osf.io] to user [krs8n]!
[__main__]  WARNING: Action skipped due to no eligible email found for user [cvenb]!
[__main__]  INFO: SUCCEEDED (2/5) = ['wbg6v', 'krs8n']
[__main__]  WARNING: SKIPPED (2/5) = ['mtf5k', 'cvenb']
[__main__]  ERROR: FAILED (1/5) = ['t47ka']
[__main__]  ERROR: ERRORED (0/5) = []
```

* Legends

```
SUCCEEDED = 0  # The email has successfully been added
SKIPPED = 1  # The email was added (to the user) before the script is run, or the eligible email is not found
FAILED = 2  # The email failed to be added since it belongs to another account
ERRORED = 3  # The email failed to be added due to unexpected exception
```

## QA Notes

N/A

## Documentation

- [ ] Product / admin notes update?

## Side Effects

* Since this script can only be run by privileged admins, email verification / spam check is not explicitly added.

## Ticket

https://openscience.atlassian.net/browse/ENG-3834

mfraezz commented on scripts/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,117 @@
+        domain_part = email_parts[1].lower()
+        if domain_part == src_domain:
+            email_to_add = f'{name_part}@{dst_domain}'
+            break
```
Implausible, but what if the user has multiple emails at the `src_domain` in question already? Should this update all of them?

cslzchen replied:
I agree. The better solution is to add every eligible email instead of the first one found and I will make the change.

If the institution only migrate one of their emails (e.g. the primary email), then we will end up with extra invalid emails. However, by adding all of them, at least will will have the correct one added.

I am drafting an email to them with other questions and will include this note. In the future, we will have another script to make the new email primary. We may not be able to decide which one to use if they have multiple.

cslzchen replied:
I need to re-do the output part since `1` -> `N`  changed how successes/failures/errors are stored and presented.

mfraezz commented on scripts/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,117 @@
+
+    # Find the eligible email to add
+    email_to_add = None
+    for email in user.emails.all():
```
Minor: Can be filtered with `user.emails.filter(address__endswith=f'@{src_domain}')`

cslzchen replied:
Done

mfraezz commented on scripts/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,117 @@
+        return
+
+    # Find all users affiliated with the given institution
+    affiliated_users = OSFUser.objects.filter(affiliated_institutions___id=institution_id)
```
Minor: Consider adding an `.exclude` subquery here to filter out users that have already been updated? Not strictly necessary, as that case results in `SKIPPED`, but may be viable way to optimize if this takes too long.

cslzchen replied:
Will add the filter. I thought the "skipped" list would be useful but probably not since it's not something we will have to manually deal later (unlike the failed ones).

cslzchen replied:
hmm, this seems not doable. It is hard to tell at this time whether a user has been updated or not since we can't just rely on whether user has an `@osf.io` email. e.g. if user has `email1@email.osf.io`, `email2@email.osf.io` and `email2@osf.io`; or if user has `email1@email.osf.io` and only `email2@osf.io`.

mfraezz replied:
It's technically possible, but may not be supported by the DjangORM, and might require `RawSQL`. Not worth the effort unless this script ends up taking significantly longer than I expect it to.

mfraezz commented on scripts/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,117 @@
+
+
+if __name__ == '__main__':
+    main()
```
Please make this a management command instead; `/scripts/` is generally deprecated and not preferred for new work that may need to be ran multiple times.

cslzchen replied:
Good catch, will move this to the management command.

cslzchen replied:
Done

mfraezz commented on scripts/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,117 @@
+    if not duplicate_user:
+        try:
+            user.emails.create(address=email_to_add)
+            user.save()
```
Is save required here? M2M changes such as `.emails.create` typically don't require saves on the related model.

cslzchen replied:
You are right! No need for `user.save()`.

cslzchen replied:
Done

mfraezz reviewed (commented):
Generally looks good, a couple questions/comments and one request. Pass complete :octocat:

cslzchen commented:
## CR Update

### Changes

- Moved from scripts to django command
- Improved query
- Supports multi-emails
- Re-worked output
- Added dry-run

### DevOps Update

```
docker-compose run --rm web python3 manage.py <institution_id>  <src_domain> <dst_domain> [--dry]
```

```
positional arguments:
  institution_id        the institution whose affiliated users' eligible email
                        is to be updated
  src_domain            the email domain that are eligible for update
  dst_domain            the email domain that is to be added

optional arguments:
  --dry                 If true, iterate through eligible users and emails but
                        don't add the email
```

### Dry-run Output

```
[osf.management.commands.update_institution_sso_email_domain]  WARNING: This is a dry-run pass.
[osf.management.commands.update_institution_sso_email_domain]  WARNING: Action aborted since email [longze@test.osf.io] exists for a different user [wbg6v]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[osf.management.commands.update_institution_sso_email_domain]  INFO: Successfully added email [roger@test.osf.io] to user [wbg6v]!
[osf.management.commands.update_institution_sso_email_domain]  INFO: Action skipped since email [naxin@test.osf.io] exists for the same user [mtf5k]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[osf.management.commands.update_institution_sso_email_domain]  INFO: Successfully added email [longze+noitailiffa@test.osf.io] to user [krs8n]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[osf.management.commands.update_institution_sso_email_domain]  INFO: Successfully added email [longze+affiliation@test.osf.io] to user [krs8n]!
[osf.management.commands.update_institution_sso_email_domain]  WARNING: Action skipped due to no eligible email found for user [cvenb]!
[osf.management.commands.update_institution_sso_email_domain]  INFO: SUCCEEDED = {'wbg6v': ['roger@test.osf.io'], 'krs8n': ['longze+noitailiffa@test.osf.io', 'longze+affiliation@test.osf.io']}
[osf.management.commands.update_institution_sso_email_domain]  INFO: SKIPPED = {'mtf5k': ['naxin@test.osf.io']}
[osf.management.commands.update_institution_sso_email_domain]  WARNING: FAILED = {'t47ka': ['longze@test.osf.io']}
[osf.management.commands.update_institution_sso_email_domain]  ERROR: ERRORED = {}
[osf.management.commands.update_institution_sso_email_domain]  WARNING: The above output is the result from a dry-run pass! SUCCEEDED ones were not actually added! ERRORED is expected to be an empty list!
```

### Actual Run Output

```
[osf.management.commands.update_institution_sso_email_domain]  WARNING: Action aborted since email [longze@test.osf.io] exists for a different user [wbg6v]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[osf.management.commands.update_institution_sso_email_domain]  INFO: Successfully added email [roger@test.osf.io] to user [wbg6v]!
[osf.management.commands.update_institution_sso_email_domain]  INFO: Action skipped since email [naxin@test.osf.io] exists for the same user [mtf5k]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[osf.management.commands.update_institution_sso_email_domain]  INFO: Successfully added email [longze+noitailiffa@test.osf.io] to user [krs8n]!
[framework.auth.core]  ERROR: OSFUser matching query does not exist.
[osf.management.commands.update_institution_sso_email_domain]  INFO: Successfully added email [longze+affiliation@test.osf.io] to user [krs8n]!
[osf.management.commands.update_institution_sso_email_domain]  WARNING: Action skipped due to no eligible email found for user [cvenb]!
[osf.management.commands.update_institution_sso_email_domain]  INFO: SUCCEEDED = {'wbg6v': ['roger@test.osf.io'], 'krs8n': ['longze+noitailiffa@test.osf.io', 'longze+affiliation@test.osf.io']}
[osf.management.commands.update_institution_sso_email_domain]  INFO: SKIPPED = {'mtf5k': ['naxin@test.osf.io']}
[osf.management.commands.update_institution_sso_email_domain]  WARNING: FAILED = {'t47ka': ['longze@test.osf.io']}
[osf.management.commands.update_institution_sso_email_domain]  ERROR: ERRORED = {}
```

cslzchen commented on osf/management/commands/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,146 @@
+        for user in affiliated_users:
+            user_result = update_email_domain(user, src_domain, dst_domain, dry_run=dry_run)
+            if user_result.get(UpdateResult.SUCCEEDED.name):
+                update_result.get(UpdateResult.SUCCEEDED.name)[user._id] = user_result.get(UpdateResult.SUCCEEDED.name)
```
We iterate the users to add the email but the output would have been very difficult to read if it is keyed by the user on the top level. Thus, covert it to be keyed by the update status first and then by the user when generating the output.

cslzchen commented on osf/management/commands/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,146 @@
+    """
+
+    # Find eligible emails to add
+    emails_to_add = []
+    user_result = {
+        UpdateResult.SUCCEEDED.name: [],
+        UpdateResult.SKIPPED.name: [],
+        UpdateResult.FAILED.name: [],
+        UpdateResult.ERRORED.name: [],
+    }
```
The tricky part about taking care of multiple emails is that one may fail, another may be skipped and the rest are successful. After trying a few options, I discovered using a map is here to categorize them is the best option.

cslzchen commented on osf/management/commands/update_institution_sso_email_domain.py:
```diff
@@ -0,0 +1,146 @@
+                logger.error(f'An unexpected error occurred when adding email [{email}] to user [{user._id}]!')
+                sentry.log_exception()
+                user_result.get(UpdateResult.ERRORED.name).append(email)
+            else:
+                logger.info(f'Successfully added email [{email}] to user [{user._id}]!')
+                user_result.get(UpdateResult.SUCCEEDED.name).append(email)
```
I decided to log "success" and update result map even it is a dry run so that you can actually see the emails to be added in the output. The differences between dry run and actual run are:

* Emails are not added
* Error list will be always be empty since unexpected exception can't happen when emails are not added
